helix-core/src/main/java/org/apache/helix/monitoring/metrics/implementation/RebalanceLatencyGauge.java

```diff
@@ -0,0 +1,94 @@
+package org.apache.helix.monitoring.metrics.implementation;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import com.codahale.metrics.Histogram;
+import com.codahale.metrics.SlidingTimeWindowArrayReservoir;
+import java.util.concurrent.TimeUnit;
+import org.apache.helix.HelixException;
+import org.apache.helix.monitoring.metrics.model.LatencyMetric;
+import org.apache.helix.monitoring.mbeans.dynamicMBeans.DynamicMetric;
+
+public class RebalanceLatencyGauge extends LatencyMetric {
+  private static final long VALUE_NOT_SET = -1;
+  private long _lastEmittedMetricValue = VALUE_NOT_SET;
+
+  /**
+   * Instantiates a new Histogram dynamic metric.
+   * @param metricName the metric name
+   */
+  public RebalanceLatencyGauge(String metricName, long slidingTimeWindow) {
+    super(metricName, new Histogram(
+        new SlidingTimeWindowArrayReservoir(slidingTimeWindow, TimeUnit.MILLISECONDS)));
+    _metricName = metricName;
+  }
+
+  /**
+   * Calling this method multiple times would simply overwrite the previous state. This is because
+   * the rebalancer could fail at any point, and we want it to recover gracefully by resetting the
+   * internal state of this metric.
+   */
+  @Override
+  public void startMeasuringLatency() {
+    reset();
+    _startTime = System.currentTimeMillis();
+  }
+
+  @Override
+  public void endMeasuringLatency() {
+    if (_endTime != 0L) {
+      throw new HelixException(
+          String.format("Needs to call startMeasuringLatency first! Metric name: %s", _metricName));
+    }
+    _endTime = System.currentTimeMillis();
+    _lastEmittedMetricValue = _endTime - _startTime;
+    updateValue(_lastEmittedMetricValue);
+  }
+
+  @Override
+  public String getMetricName() {
+    return _metricName;
+  }
+
+  @Override
+  public void reset() {
```

Comment:
Why this is public?
Consider the following call sequence:
1. startMeasuringLatency()
2. reset()
3. endMeasuringLatency(), then you record a huge number as the latency.

Reply:
Would this work?

```
  @Override
  public void endMeasuringLatency() {
    if (_startTime == 0L || _endTime != 0L) {
      LOG.error(
          "Needs to call startMeasuringLatency first! Ignoring and resetting the metric. Metric name: {}",
          _metricName);
      reset();
    }
    _endTime = System.currentTimeMillis();
    _lastEmittedMetricValue = _endTime - _startTime;
    updateValue(_lastEmittedMetricValue);
  }
```

Reply:
No, in this case, instead of resetting and continue the logic, I would prefer reset and skip the updateValue().

How about:
1. reset set both value to be -1.
2. if _startTime is -1, ignore the endMeasuringLatency calls.
3. After update value, force reset.

Reply:
Also need to check _endTime != VALUE_NOT_SET